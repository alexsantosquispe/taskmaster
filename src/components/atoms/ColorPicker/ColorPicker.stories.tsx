import type { Meta, StoryObj } from '@storybook/react';

import { Wrapper } from '@/utils/wrappers.utils';
import { ColorPicker } from './ColorPicker';

const meta: Meta<typeof ColorPicker> = {
  title: 'Atoms/ColorPicker',
  component: ColorPicker,
  parameters: {
    layout: 'centered'
  }
};

export default meta;

type Story = StoryObj<typeof ColorPicker>;

export const Default: Story = () => {
  return (
    <div className="flex gap-4">
      <Wrapper>
        <ColorPicker label="Color picker" />
      </Wrapper>
      <Wrapper>
        <ColorPicker label="Picker disabled" isDisabled={true} />
      </Wrapper>
    </div>
  );
};

Default.args = {};
