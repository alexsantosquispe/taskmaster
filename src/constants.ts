import { PRIORITY_TYPES } from './models/types';

export const THEME_KEY = 'theme';

export const PRIORITY_COLORS = {
  [PRIORITY_TYPES.LOW]:
    'bg-amber-100 text-yellow-800 border border-yellow-800 dark:bg-yellow-600/30 dark:border-yellow-400 dark:text-yellow-400',
  [PRIORITY_TYPES.MEDIUM]:
    'bg-orange-100 text-amber-800 border border-amber-800 dark:bg-orange-600/20 dark:border-orange-400 dark:text-orange-400',
  [PRIORITY_TYPES.HIGH]:
    'bg-red-100 text-rose-700 border border-rose-700 dark:bg-red-600/20 dark:border-red-400 dark:text-red-400'
};

export const currentUser = {
  name: 'Alex Santos',
  email: 'alex.santos456@gmail.com',
  avatarUrl: '/img/avatar.webp'
};
