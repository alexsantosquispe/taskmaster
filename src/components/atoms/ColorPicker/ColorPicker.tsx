import { COLORS, DEFAULT_COLOR } from '@/constants';
import type {
  ColorPickerProps,
  ColorPickerWithControllerProps,
  Option
} from '@/models/types';

import cn from 'clsx';
import { useState } from 'react';
import { Controller } from 'react-hook-form';
import { twMerge } from 'tailwind-merge';
import { IconButton } from '../IconButton/IconButton';

export const ColorPicker = ({
  label,
  value,
  isRequired = false,
  isDisabled = false,
  onChangeColor,
  className
}: ColorPickerProps) => {
  const [currentColor, setCurrentColor] = useState<Option>(
    value || DEFAULT_COLOR
  );
  const [isPalletColorOpen, setIsPalletColorOpen] = useState(false);

  const togglePalletColor = () => {
    setIsPalletColorOpen(!isPalletColorOpen);
  };

  const onSelectColor = (color: Option) => {
    setCurrentColor(color);
    if (typeof onChangeColor === 'function') {
      onChangeColor(color);
    }
    setIsPalletColorOpen(false);
  };

  return (
    <div className="relative flex flex-col">
      {label && (
        <label className="mb-0.5 text-sm">
          {label}
          {isRequired && <span className="text-rose-600">&nbsp;*</span>}
        </label>
      )}
      <IconButton
        ariaLabel="Color picker button"
        icon={
          <div
            className={twMerge('h-5 w-5 rounded-full', currentColor.label)}
          />
        }
        onClick={togglePalletColor}
        isDisable={isDisabled}
        className={twMerge(
          'border border-neutral-200 dark:border-white/20',
          className
        )}
        tooltipMessage="Select color"
      />
      {isPalletColorOpen && (
        <div
          data-testid="color-pallet"
          className="dark:bg-primary absolute top-full right-0 mt-1 w-[12.25rem] rounded-lg border border-neutral-200 bg-white p-2 shadow-md dark:border-white/20 dark:shadow-none"
        >
          <div className="flex flex-wrap gap-4">
            {COLORS.map((color, index) => {
              const isSelected = color.value === currentColor.value;
              return (
                <button
                  key={`${color}-${index}`}
                  type="button"
                  aria-label={`Color code ${color.label}`}
                  onClick={() => onSelectColor(color)}
                  className={twMerge(
                    'hover:ring-accent dark:hover:ring-accent-dark h-4 w-4 rounded-full hover:cursor-pointer hover:ring-2 hover:ring-offset-2',
                    color.label,
                    cn({
                      'ring-accent dark:ring-accent-dark ring-2 ring-offset-2':
                        isSelected
                    })
                  )}
                />
              );
            })}
          </div>
        </div>
      )}
    </div>
  );
};

export const ColorPickerWithController = ({
  name,
  control,
  ...props
}: ColorPickerWithControllerProps) => {
  return (
    <Controller
      control={control}
      name={name}
      render={({ field: { onChange, value } }) => (
        <ColorPicker {...props} onChangeColor={onChange} value={value} />
      )}
    />
  );
};
