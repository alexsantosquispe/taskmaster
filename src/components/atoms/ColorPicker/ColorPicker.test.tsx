import { fireEvent, render, screen } from '@testing-library/react';
import { ColorPicker, ColorPickerWithController } from './ColorPicker';

import { COLORS } from '@/constants';
import type { ColorPickerProps } from '@/models/types';
import { useForm } from 'react-hook-form';

const ColorPickerWrapperController = (props: ColorPickerProps) => {
  const { control } = useForm();
  return (
    <ColorPickerWithController {...props} name="color" control={control} />
  );
};

describe('ColorPicker', () => {
  const onChangeColorMock = jest.fn();
  const props = {
    label: 'Color picker',
    onChangeColor: onChangeColorMock
  };

  describe('styles', () => {
    it('should render the color picker correctly', () => {
      const component = render(<ColorPicker {...props} />);

      expect(component).toMatchSnapshot();
    });

    it('should render the color picker with controller correctly', () => {
      const component = render(<ColorPickerWrapperController {...props} />);

      expect(component).toMatchSnapshot();
    });
  });

  describe('behavior', () => {
    const numberOfColorButtons = COLORS.length + 1;
    const redColor = COLORS[5];

    it('should open the pallet color available and pick one', () => {
      render(<ColorPicker {...props} />);

      const button = screen.getByRole('button');

      fireEvent.click(button);

      const colorPallet = screen.getByTestId('color-pallet');

      expect(colorPallet).toBeInTheDocument();

      expect(screen.getAllByRole('button')).toHaveLength(numberOfColorButtons);

      const colorToSelect = screen.getByRole('button', {
        name: `Color code ${redColor.label}`
      });

      fireEvent.click(colorToSelect);

      expect(onChangeColorMock).toHaveBeenCalledTimes(1);

      expect(onChangeColorMock).toHaveBeenCalledWith(redColor);
    });
  });
});
