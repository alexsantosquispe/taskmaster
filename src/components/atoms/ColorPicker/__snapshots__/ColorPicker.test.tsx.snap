// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ColorPicker styles should render the color picker correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="relative flex flex-col"
      >
        <label
          class="mb-0.5 text-sm"
        >
          Color picker
        </label>
        <div
          class="relative flex items-center justify-center"
          data-testid="icon-button"
        >
          <button
            aria-label="Color picker button"
            class="text-primary flex w-fit items-center justify-center rounded-lg p-2 transition-colors duration-150 ease-in-out dark:text-white/70 hover:text-accent dark:hover:text-accent-dark hover:cursor-pointer hover:bg-neutral-100 dark:hover:bg-white/15 border border-neutral-200 dark:border-white/20"
            type="button"
          >
            <div
              class="h-5 w-5 rounded-full bg-blue-600"
            />
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="relative flex flex-col"
    >
      <label
        class="mb-0.5 text-sm"
      >
        Color picker
      </label>
      <div
        class="relative flex items-center justify-center"
        data-testid="icon-button"
      >
        <button
          aria-label="Color picker button"
          class="text-primary flex w-fit items-center justify-center rounded-lg p-2 transition-colors duration-150 ease-in-out dark:text-white/70 hover:text-accent dark:hover:text-accent-dark hover:cursor-pointer hover:bg-neutral-100 dark:hover:bg-white/15 border border-neutral-200 dark:border-white/20"
          type="button"
        >
          <div
            class="h-5 w-5 rounded-full bg-blue-600"
          />
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`ColorPicker styles should render the color picker with controller correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="relative flex flex-col"
      >
        <label
          class="mb-0.5 text-sm"
        >
          Color picker
        </label>
        <div
          class="relative flex items-center justify-center"
          data-testid="icon-button"
        >
          <button
            aria-label="Color picker button"
            class="text-primary flex w-fit items-center justify-center rounded-lg p-2 transition-colors duration-150 ease-in-out dark:text-white/70 hover:text-accent dark:hover:text-accent-dark hover:cursor-pointer hover:bg-neutral-100 dark:hover:bg-white/15 border border-neutral-200 dark:border-white/20"
            type="button"
          >
            <div
              class="h-5 w-5 rounded-full bg-blue-600"
            />
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="relative flex flex-col"
    >
      <label
        class="mb-0.5 text-sm"
      >
        Color picker
      </label>
      <div
        class="relative flex items-center justify-center"
        data-testid="icon-button"
      >
        <button
          aria-label="Color picker button"
          class="text-primary flex w-fit items-center justify-center rounded-lg p-2 transition-colors duration-150 ease-in-out dark:text-white/70 hover:text-accent dark:hover:text-accent-dark hover:cursor-pointer hover:bg-neutral-100 dark:hover:bg-white/15 border border-neutral-200 dark:border-white/20"
          type="button"
        >
          <div
            class="h-5 w-5 rounded-full bg-blue-600"
          />
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
